src/olmo_core/train/callbacks/batch_size_scheduler.py

```diff
@@ -0,0 +1,171 @@
+                    assert target_batch_size % self.current_batch_size == 0
+                    # When increasing the batch size, need to wait until 'batches_processed' is divisible
+                    # by the factor that batch size is being increased by.
+                    ratio = target_batch_size // self.current_batch_size
+                    if self.trainer.data_loader.batches_processed % ratio == 0:
+                        self._update_batch_size_and_lr(target_batch_size)
+                elif target_batch_size < self.current_batch_size:
+                    assert self.current_batch_size % target_batch_size == 0
+                    self._update_batch_size_and_lr(target_batch_size)
+
+                break
+
+    def _update_batch_size_and_lr(self, batch_size: int):
+        log.info(f"Changing global batch size to {batch_size:,d}...")
+        ratio = batch_size / self.current_batch_size
+        lr_adjustment_factor = math.sqrt(ratio)
+        self.trainer.data_loader.global_batch_size = batch_size
+
+        optimizers: Optional[List[torch.optim.Optimizer]] = None
+        scheduler: Optional[Scheduler] = None
+        if isinstance(self.trainer.train_module, TransformerTrainModule):
+            optimizers = [self.trainer.train_module.optim]
+            scheduler = self.trainer.train_module.scheduler
+        elif isinstance(self.trainer.train_module, TransformerPipelineTrainModule):
+            optimizers = self.trainer.train_module.optimizers
+            scheduler = self.trainer.train_module.scheduler
+
+        if not optimizers:
+            log.warning(
+                f"Unable to adjust learning rate for {self.trainer.train_module.__class__.__name__} train module class"
+            )
+            return
+
+        for optim in optimizers:
+            if not isinstance(optim, (torch.optim.Adam, torch.optim.AdamW, SkipStepAdamW)):
+                log.warning(
+                    f"Unable to adjust learning rate for {optim.__class__.__name__} optimizer"
+                )
+                return
+
+        log.info(
+            f"Adjusting base learning rate by a factor of {lr_adjustment_factor:.4f} = sqrt({ratio})"
+        )
+        for optim_idx, optim in enumerate(optimizers):
+            for group_idx, group in enumerate(optim.param_groups):
+                lr_field = LR_FIELD if scheduler is None else scheduler.lr_field
+                initial_lr_field = (
+                    INITIAL_LR_FIELD if scheduler is None else scheduler.initial_lr_field
+                )
+
+                if group.get(initial_lr_field) is None:
+                    group[initial_lr_field] = group[lr_field]
+
+                group[initial_lr_field] *= lr_adjustment_factor
+
+                # Only set the actual LR if there's no scheduler. Schedulers update the LR based
+                # on the initial LR.
+                if scheduler is None:
+                    group[lr_field] = group[initial_lr_field]
```

Comment:
LR schedulers don't overwrite this every step?

Reply:
They overwrite "lr" but not "initial_lr". So we only touch "lr" when there's no scheduler.